File: java/ZipDu.java
import java.io.*;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.Path;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;
import java.util.zip.ZipInputStream;

import javax.servlet.MultipartConfigElement;
import javax.servlet.http.Part;

import static spark.Spark.*;
import spark.utils.IOUtils;
import com.google.gson.*; 

public class ZipDu {
  public static Gson GSON = new Gson();
  public static String uploadsFolder = "uploads";

  public static void main(String[] args) {
    // check if uploads directory exists and exit if not
    Path uploadsDirectory = Paths.get(uploadsFolder);
    if (!Files.exists(uploadsDirectory)) {
      System.err.println("Could not find the `uploads` directory in the folder you executed zipdu in. Exitting.");
      System.exit(1);
    }

    // health route
    get("/health", (req, res) -> {
      res.type("application/json");

      Map<String, Object> resMap = new HashMap<String, Object>();
      resMap.put("ok", true);
      return GSON.toJson(resMap);
    });

    // zipstats route
    post("/zipstats", (req, res) -> {
      res.type("application/json");

      req.attribute("org.eclipse.jetty.multipartConfig", new MultipartConfigElement(System.getProperty("java.io.tmpdir")));
      Part filePart = req.raw().getPart("file");
      if (filePart == null) {
        res.status(400);
        return "";
      }

      ZipInputStream zis = null;
      try {
        zis = new ZipInputStream(filePart.getInputStream());
      } catch(FileNotFoundException e) {}
      if (zis == null) {
        res.status(400);
        return "";
      }

      // create directory inside the uploads directory
      String directoryName = UUID.randomUUID().toString();
      Path directoryPath = Paths.get(uploadsFolder, directoryName);
      File outputDirectory = new File(directoryPath.toString());
      if (outputDirectory.exists()) {
        try { zis.close(); } catch (IOException e) {}
        res.status(400);
        return "";
      }
      if (!outputDirectory.mkdir()) {
        try { zis.close(); } catch (IOException e) {}
        res.status(400);
        return "";
      }

      // actual unzip operation
      int numberOfFiles = 0;
      int totalSize = 0;
      try {
        byte[] buffer = new byte[2048];
        ZipEntry zipEntry = zis.getNextEntry();
        while (zipEntry != null) {
          File destinationFile = new File(directoryPath.toString(), zipEntry.getName());
          if (zipEntry.isDirectory()) {
            destinationFile.mkdirs();
            zipEntry = zis.getNextEntry();
            continue;
          }
          FileOutputStream outStream = new FileOutputStream(destinationFile);
          int written;
          while ((written = zis.read(buffer)) > 0) {
            outStream.write(buffer, 0, written);
            totalSize += written;
          }
          outStream.close();
          numberOfFiles += 1;

          zipEntry = zis.getNextEntry();
        }
      } catch (IOException e) {
        res.status(400);
        return "";
      } finally {
        try {
          zis.closeEntry();
          zis.close();
        } catch (IOException e) {
          res.status(400);
          return e.toString();
        }
      }

      // write response
      Map<String, Object> resMap = new HashMap<String, Object>();
      resMap.put("uuid", directoryName);
      resMap.put("totalSize", totalSize);
      resMap.put("numberOfFiles", numberOfFiles);
      return GSON.toJson(resMap);
    });

    // notFound route
    notFound((req, res) -> {
      res.status(404);
      return "";
    });
  }
}

File: scala/ZipDu.scala
import java.io._
import java.net._

import java.nio.file.{Files, Paths, Path}

import java.util.{Map, UUID}
import java.util.zip.{ZipEntry, ZipFile, ZipInputStream}

import javax.servlet.MultipartConfigElement
import javax.servlet.http.Part

import scala.collection.mutable.HashMap
import scala.collection.JavaConverters._
import scala.util.control.Breaks._

import spark.Spark._
import spark.{Request, Response, Route}
import spark.utils.IOUtils
import com.google.gson._

object ZipDu {
  val GSON = new Gson()
  val uploadsFolder = "uploads"

  def main(args: Array[String]) = {
    // check if uploads directory exists and exit if not
    val uploadsDirectory = Paths.get(uploadsFolder)
    if (!Files.exists(uploadsDirectory)) {
      System.err.println("Could not find the `uploads` directory in the folder you executed zipdu in. Exitting.")
      System.exit(1)
    }

    // health route
    get("/health", new Route {
      override def handle(request: Request, response: Response): String = {
        val resMap = HashMap("ok" -> true)
        return GSON.toJson(resMap)
      }
    })

    // zipstats route
    post("/zipstats", new Route {
      override def handle(req: Request, res: Response): String = {
          req.attribute("org.eclipse.jetty.multipartConfig", new MultipartConfigElement(System.getProperty("java.io.tmpdir")))
          val filePart = req.raw().getPart("file")
          if (filePart == null) {
            res.status(400)
            return ""
          }

          var zis : ZipInputStream = null
          try {
            zis = new ZipInputStream(filePart.getInputStream())
          } catch {
            case e: FileNotFoundException => {}
          }
          if (zis == null) {
            res.status(400)
            return ""
          }

          // create directory inside the uploads directory
          val directoryName = UUID.randomUUID().toString()
          val directoryPath = Paths.get(uploadsFolder, directoryName)
          val outputDirectory = new File(directoryPath.toString())
          if (outputDirectory.exists()) {
            try { zis.close() } catch { case e: IOException => {} }
            res.status(400)
            return ""
          }
          if (!outputDirectory.mkdir()) {
            try { zis.close() } catch { case e: IOException => {} }
            res.status(400)
            return ""
          }

          // actual unzip operation
          var numberOfFiles: Int = 0
          var totalSize: Int = 0
          try {
            var buffer = new Array[Byte](2048)
            var zipEntry = zis.getNextEntry()
            while (zipEntry != null) {
              breakable {
                val destinationFile = new File(directoryPath.toString(), zipEntry.getName())
                if (zipEntry.isDirectory()) {
                  destinationFile.mkdirs()
                  zipEntry = zis.getNextEntry()
                  break
                }
                val outStream = new FileOutputStream(destinationFile)
                var written: Int = zis.read(buffer)
                while (written > 0) {
                  outStream.write(buffer, 0, written)
                  totalSize += written
                  written = zis.read(buffer)
                }
                outStream.close()
                numberOfFiles += 1

                zipEntry = zis.getNextEntry()
              }
            }
          } catch {
            case e: IOException => {
              res.status(400)
              return ""
            }
          } finally {
            try {
              zis.closeEntry()
              zis.close()
            } catch {
              case e: IOException => {
                res.status(400)
                return e.toString()
              }
            }
          }

          // write response
          val resMap = HashMap(
            "uuid" -> directoryName,
            "totalSize" -> totalSize,
            "numberOfFiles" -> numberOfFiles)
          return GSON.toJson(resMap.asJava)
       }
    })

    notFound(new Route {
      override def handle(req: Request, res: Response): String = {
        res.status(404)
        return ""
      }
    })
  }
}
